### PR TITLE
chore: Add a GitHub action that runs unit tests with -race to CI build (#4774)

### PR DESCRIPTION
**Pull request by jgwest**
See [parent issue](https://github.com/argoproj/argo-cd/issues/4774) for the full details of the PR.

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

Signed-off-by: Jonathan West <jonwest@redhat.com>

### CODE REVIEW

**Review by jannfis — CHANGES_REQUESTED** [review]
Awesome, I like the idea to leverage race detection in our CI builds. Also, I like the implementation of segregating our technical debt (or existing bugs...) to be excluded, so it gives us time to fix them while preventing new race conditions to sneak in.

If I understood it correctly, we'll be progressively fix the existing race conditions and then move them to the "normal" unit tests without the `!race` pragma. Is that a correct understanding?

One last question: Will we still be able to run race tests explicitly (i.e. using `go test -race -run '...'`) with those build pragmas added to the files?

I have a couple of minor comments, please see inline.

PS: Thanks (again) for the very detailed problem description and writing down your train of thoughts in the related issue, can't stress how much appreciated this is!

File: .github/workflows/ci-build.yaml [hunk 1]
```diff
@@ -132,6 +132,66 @@ jobs:
...
+      - name: Download and vendor all required packages
+        run: |
+          go mod download
+      - name: Run all unit tests
+        run: make test-race-local
+      - name: Generate code coverage artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: code-coverage
+          path: coverage.out
```
[the remark this answers is not in the document]
Comment (jgwest) [reply .github/workflows/ci-build.yaml]: Aha, good catch, I have removed this step.

File: .github/workflows/ci-build.yaml [hunk 2]
```diff
@@ -132,6 +132,66 @@ jobs:
...
+      - name: Generate test results artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: test-results
```
[the remark this answers is not in the document]
Comment (jgwest) [reply .github/workflows/ci-build.yaml]: Good thinking, still new to GitHub actions so I had assumed each job had a different upload scope, but obviously not. Fixed to `race-results`.

**Comment by jgwest** [comment]
> If I understood it correctly, we'll be progressively fix the existing race conditions and then move them to the "normal" unit tests without the !race pragma. Is that a correct understanding?

This is exactly correct... we don't necessarily want to wait for all possible issues to be fixed before we enable it, but we can enable this now to prevent further data races from being introduced to the tests. Once that's done, we can progressively fix the remaining issues (which are clearly visible just by looking for the `_norace_` files and/or the build pragma). And of course, as mentioned in the original issue, I've gone and squashed a bunch of existing races in order to get us to where we can enable this.


> One last question: Will we still be able to run race tests explicitly (i.e. using go test -race -run '...') with those build pragmas added to the files?

Afraid not, or at least I'm not aware of a way to tell `go test` to explicitly override the pragma. If you attempt to individually run one of these with `-race`, as in your example, you will get a `no test files` message from `go test`.

In my local development environment I've just been removing the `// +build !race` tag whenever I need to test an individual test, and I'm sure one could construct a recursive `sed` to temporarily remove them if necessary (which is imperfect, but works).

(@jannfis)

**Comment by jannfis** [comment]
> > One last question: Will we still be able to run race tests explicitly (i.e. using go test -race -run '...') with those build pragmas added to the files?
> 
> Afraid not, or at least I'm not aware of a way to tell `go test` to explicitly override the pragma. If you attempt to individually run one of these with `-race`, as in your example, you will get a `no test files` message from `go test`.
> 
> In my local development environment I've just been removing the `// +build !race` tag whenever I need to test an individual test, and I'm sure one could construct a recursive `sed` to temporarily remove them if necessary (which is imperfect, but works).

That's fair enough from my pov. It's obvious from the outside that the contained tests are run with `!race` pragma, and for manual re-testing, removing the pragma locally by hand is not much of a hassle. I think we won't need some `sed`-script/command to do that.

We should document the `_norace_` tests in the contribution guide, however (tho not necessarily with this PR). I hope I can finish setting up the contribution documentation repository this weekend and migrate existing documentation. Then updating the docs would not be that much of a hassle as it is right now.

**Review by jannfis — APPROVED** [review]
LGTM

**Comment by jannfis** [comment]
@alexmt @jessesuen FYI

**Pull request merged**